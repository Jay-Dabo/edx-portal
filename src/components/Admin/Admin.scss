.row.equal-col-height {
    display: flex;
    flex-wrap: wrap;

    > [class*="col-"] {
        display: flex;
        flex-direction: column;
    }

    .card {
        flex: 1;
    }
}
