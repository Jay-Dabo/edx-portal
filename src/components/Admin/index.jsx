import React from 'react';
import PropTypes from 'prop-types';
import Helmet from 'react-helmet';
import { StatusAlert, MailtoLink } from '@edx/paragon';

import H1 from '../../components/H1';
import H2 from '../../components/H2';
import NumberCard from '../../components/NumberCard';
import CourseEnrollmentsTable from '../../containers/CourseEnrollmentsTable';

import './Admin.scss';

const StatusMessage = props => (
  <StatusAlert
    alertType={props.alertType}
    dismissible={false}
    dialog={props.message}
    open
  />
);

class Admin extends React.Component {
  componentDidMount() {
    // TODO: enterprise uuid will be retrieved from data we get back about user
    // during authentication.
    const enterpriseId = 'ee5e6b3a-069a-4947-bb8d-d2dbc323396c';
    this.props.getDashboardAnalytics(enterpriseId);
  }

  hasAnalyticsData() {
    const {
      activeLearners,
      courseCompletions,
      enrolledLearners,
    } = this.props;
    let hasAnalyticsData = false;

    [activeLearners, courseCompletions, enrolledLearners].forEach((item) => {
      hasAnalyticsData = item !== null || hasAnalyticsData;
    });

    return hasAnalyticsData;
  }

  renderErrorMessage() {
    const message = `There was a problem fetching dashboard data: ${this.props.error.message}`;
    return <StatusMessage alertType="danger" message={message} />;
  }

  renderLoadingMessage() {
    return (
      <div className="loading">
        <p>Loading...</p>
      </div>
    );
  }

  render() {
    const {
      activeLearners,
      courseCompletions,
      enrolledLearners,
      error,
      loading,
    } = this.props;

    return (
      <div className="container">
        <Helmet>
          <title>edX Learner and Progress Report</title>
        </Helmet>
        <div className="row">
          <div className="col">
            <H1>edX Learner and Progress Report</H1>
          </div>
        </div>
        <div className="row">
          <div className="col">
            <H2>Overview</H2>
            {error && this.renderErrorMessage()}
            {loading && this.renderLoadingMessage()}
            {!loading && !error && this.hasAnalyticsData() &&
              <div className="row mt-3 equal-col-height">
                <div className="col-xs-12 col-md-6 col-xl-3 mb-3">
                  <NumberCard
                    title="X"
                    description="total number of learners registered"
                    iconClassName="fa fa-users"
                  />
                </div>
                <div className="col-xs-12 col-md-6 col-xl-3 mb-3">
                  <NumberCard
                    title={enrolledLearners}
                    description="learners enrolled in at least one course"
                    iconClassName="fa fa-check"
                  />
                </div>
                <div className="col-xs-12 col-md-6 col-xl-3 mb-3">
                  <NumberCard
                    title={activeLearners.past_week}
                    description="active learners in the past week"
                    iconClassName="fa fa-eye"
                  />
                </div>
                <div className="col-xs-12 col-md-6 col-xl-3 mb-3">
                  <NumberCard
                    title={courseCompletions}
                    description="course completions"
                    iconClassName="fa fa-trophy"
                  />
                </div>
              </div>
            }
          </div>
        </div>
        <div className="row mt-4">
          <div className="col">
            <H2>Full Report</H2>
            <CourseEnrollmentsTable className="mt-3" />
          </div>
        </div>
        <div className="row mt-4">
          <div className="col">
            <p>
              For more information, contact edX Enterprise Support at <MailtoLink to="enterprise@edx.org" content=" enterprise@edx.org" />.
            </p>
          </div>
        </div>
      </div>
    );
  }
}

Admin.defaultProps = {
  error: null,
  loading: false,
  courseCompletions: null,
  activeLearners: null,
  enrolledLearners: null,
};

Admin.propTypes = {
  getDashboardAnalytics: PropTypes.func.isRequired,
  activeLearners: PropTypes.shape({
    past_week: PropTypes.number,
    past_month: PropTypes.number,
  }),
  enrolledLearners: PropTypes.number,
  courseCompletions: PropTypes.number,
  loading: PropTypes.bool,
  error: PropTypes.instanceOf(Error),
};

StatusMessage.propTypes = {
  message: PropTypes.string.isRequired,
  alertType: PropTypes.string.isRequired,
};

export default Admin;
