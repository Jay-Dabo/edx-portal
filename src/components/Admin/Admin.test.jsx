import React from 'react';
import PropTypes from 'prop-types';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import Admin from './index';

const mockStore = configureMockStore([thunk]);

class ContextProvider extends React.Component {
  static childContextTypes = {
    store: PropTypes.object.isRequired,
    courseEnrollments: PropTypes.shape({
      enrollments: PropTypes.object,
    }),
  }

  getChildContext = () => ({
    store: mockStore({
      courseEnrollments: {
        enrollments: {},
      },
    }),
  })

  render() {
    return this.props.children;
  }
}

ContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

const AdminWrapper = props => (
  <MemoryRouter>
    <ContextProvider>
      <Admin
        getDashboardAnalytics={() => {}}
        {...props}
      />
    </ContextProvider>
  </MemoryRouter>
);

describe('<Admin />', () => {
  describe('renders correctly', () => {
    it('calls getDashboardAnalytics prop', () => {
      const mockGetDashboardAnalytics = jest.fn();
      const tree = renderer
        .create((
          AdminWrapper({
            getDashboardAnalytics: mockGetDashboardAnalytics,
          })
        ))
        .toJSON();
      expect(mockGetDashboardAnalytics).toHaveBeenCalled();
      expect(tree).toMatchSnapshot();
    });

    it('with dashboard analytics data', () => {
      const tree = renderer
        .create((
          AdminWrapper({
            activeLearners: {
              past_week: 1,
              past_month: 1,
            },
            enrolledLearners: 1,
            courseCompletions: 1,
          })
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('with error state', () => {
      const tree = renderer
        .create((
          AdminWrapper({
            error: Error('Network Error'),
          })
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('with loading state', () => {
      const tree = renderer
        .create((
          AdminWrapper({
            loading: true,
          })
        ))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
