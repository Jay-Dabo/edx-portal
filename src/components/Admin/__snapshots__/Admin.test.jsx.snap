// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Admin /> renders correctly calls getDashboardAnalytics prop 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <h1>
        edX Learner and Progress Report
      </h1>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <h2>
        Overview
      </h2>
    </div>
  </div>
  <div
    className="row mt-4"
  >
    <div
      className="col"
    >
      <h2>
        Full Report
      </h2>
      <div>
        <div
          className="alert fade alert-warning show"
          hidden={false}
          role="alert"
        >
          <div
            className="alert-dialog"
          >
            There are no course enrollments.
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="row mt-4"
  >
    <div
      className="col"
    >
      <p>
        For more information, contact edX Enterprise Support at 
        <a
          href="mailto:enterprise@edx.org"
          onClick={null}
          target="_self"
        >
           enterprise@edx.org
        </a>
        .
      </p>
    </div>
  </div>
</div>
`;

exports[`<Admin /> renders correctly with dashboard analytics data 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <h1>
        edX Learner and Progress Report
      </h1>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <h2>
        Overview
      </h2>
      <div
        className="row mt-3 equal-col-height"
      >
        <div
          className="col-xs-12 col-md-6 col-xl-3 mb-3"
        >
          <div
            className="card"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title d-flex align-items-center justify-content-between"
              >
                <span
                  className="title"
                >
                  X
                </span>
                <div>
                  <span
                    aria-hidden={true}
                    className="fa fa-users d-flex align-items-center justify-content-center"
                    id="Icon1"
                  />
                </div>
              </h5>
              <p
                className="card-text"
              >
                total number of learners registered
              </p>
            </div>
          </div>
        </div>
        <div
          className="col-xs-12 col-md-6 col-xl-3 mb-3"
        >
          <div
            className="card"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title d-flex align-items-center justify-content-between"
              >
                <span
                  className="title"
                >
                  1
                </span>
                <div>
                  <span
                    aria-hidden={true}
                    className="fa fa-check d-flex align-items-center justify-content-center"
                    id="Icon1"
                  />
                </div>
              </h5>
              <p
                className="card-text"
              >
                learners enrolled in at least one course
              </p>
            </div>
          </div>
        </div>
        <div
          className="col-xs-12 col-md-6 col-xl-3 mb-3"
        >
          <div
            className="card"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title d-flex align-items-center justify-content-between"
              >
                <span
                  className="title"
                >
                  1
                </span>
                <div>
                  <span
                    aria-hidden={true}
                    className="fa fa-eye d-flex align-items-center justify-content-center"
                    id="Icon1"
                  />
                </div>
              </h5>
              <p
                className="card-text"
              >
                active learners in the past week
              </p>
            </div>
          </div>
        </div>
        <div
          className="col-xs-12 col-md-6 col-xl-3 mb-3"
        >
          <div
            className="card"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title d-flex align-items-center justify-content-between"
              >
                <span
                  className="title"
                >
                  1
                </span>
                <div>
                  <span
                    aria-hidden={true}
                    className="fa fa-trophy d-flex align-items-center justify-content-center"
                    id="Icon1"
                  />
                </div>
              </h5>
              <p
                className="card-text"
              >
                course completions
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="row mt-4"
  >
    <div
      className="col"
    >
      <h2>
        Full Report
      </h2>
      <div>
        <div
          className="alert fade alert-warning show"
          hidden={false}
          role="alert"
        >
          <div
            className="alert-dialog"
          >
            There are no course enrollments.
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="row mt-4"
  >
    <div
      className="col"
    >
      <p>
        For more information, contact edX Enterprise Support at 
        <a
          href="mailto:enterprise@edx.org"
          onClick={null}
          target="_self"
        >
           enterprise@edx.org
        </a>
        .
      </p>
    </div>
  </div>
</div>
`;

exports[`<Admin /> renders correctly with error state 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <h1>
        edX Learner and Progress Report
      </h1>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <h2>
        Overview
      </h2>
      <div
        className="alert fade alert-danger show"
        hidden={false}
        role="alert"
      >
        <div
          className="alert-dialog"
        >
          There was a problem fetching dashboard data: Network Error
        </div>
      </div>
    </div>
  </div>
  <div
    className="row mt-4"
  >
    <div
      className="col"
    >
      <h2>
        Full Report
      </h2>
      <div>
        <div
          className="alert fade alert-warning show"
          hidden={false}
          role="alert"
        >
          <div
            className="alert-dialog"
          >
            There are no course enrollments.
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="row mt-4"
  >
    <div
      className="col"
    >
      <p>
        For more information, contact edX Enterprise Support at 
        <a
          href="mailto:enterprise@edx.org"
          onClick={null}
          target="_self"
        >
           enterprise@edx.org
        </a>
        .
      </p>
    </div>
  </div>
</div>
`;

exports[`<Admin /> renders correctly with loading state 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <h1>
        edX Learner and Progress Report
      </h1>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <h2>
        Overview
      </h2>
      <div
        className="loading"
      >
        <p>
          Loading...
        </p>
      </div>
    </div>
  </div>
  <div
    className="row mt-4"
  >
    <div
      className="col"
    >
      <h2>
        Full Report
      </h2>
      <div>
        <div
          className="alert fade alert-warning show"
          hidden={false}
          role="alert"
        >
          <div
            className="alert-dialog"
          >
            There are no course enrollments.
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="row mt-4"
  >
    <div
      className="col"
    >
      <p>
        For more information, contact edX Enterprise Support at 
        <a
          href="mailto:enterprise@edx.org"
          onClick={null}
          target="_self"
        >
           enterprise@edx.org
        </a>
        .
      </p>
    </div>
  </div>
</div>
`;
